nixos/modules/services/monitoring/osquery.nix

```diff
@@ -0,0 +1,114 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.services.osquery;
+  logsDirectoryPrefix = "/var/log/";
+  stateDirectoryPrefix = "/var/lib/";
+  runtimeDir = "/run/osquery";
+  dirname = path: with lib.strings; with lib.lists; concatStringsSep "/"
+    (init (splitString "/" (normalizePath path)));
+
+  # conf is the osquery configuration file used when the --config_plugin=filesystem.
+  # filesystem is the osquery default value for the config_plugin flag.
+  conf = pkgs.writeText "osquery.conf" (builtins.toJSON cfg.settings);
+
+  # flagfile is the file containing osquery command line flags to be
+  # provided to the application using the special --flagfile option.
+  flagfile = pkgs.writeText "osquery.flags"
+    (concatStringsSep "\n"
+      (mapAttrsToList (name: value: "--${name}=${value}")
+        # Use the conf derivation if not otherwise specified.
+        ({ config_path = conf; } // cfg.flags)));
+
+  osqueryi = pkgs.runCommand "osqueryi" { nativeBuildInputs = [ pkgs.makeWrapper ]; } ''
+    mkdir -p $out/bin
+    ln -s ${pkgs.osquery}/bin/osqueryi $out/bin/osqueryi
+    wrapProgram $out/bin/osqueryi \
+      --add-flags "--flagfile ${flagfile}"
+    '';
+in
+{
+  options.services.osquery = {
+    enable = mkEnableOption (mdDoc "osqueryd daemon");
+
+    settings = mkOption {
+      default = { };
+      description = mdDoc ''
+        Configuration to be written to the osqueryd JSON configuration file.
+        To understand the configuration format, refer to https://osquery.readthedocs.io/en/stable/deployment/configuration/#configuration-components.
+      '';
+      example = {
+        options.utc = false;
+      };
+      type = types.attrs;
+    };
+
+    flags = mkOption {
+      default = { };
+      description = mdDoc ''
+        Attribute set of flag names and values to be written to the osqueryd flagfile.
+        For more information, refer to https://osquery.readthedocs.io/en/stable/installation/cli-flags.
+      '';
+      example = {
+        config_refresh = "10";
+      };
+      type = with types;
+        let
+          pathWithPrefix = prefix: mkOptionType {
+            name = "path with prefix";
+            description = "path with prefix \"${prefix}\"";
+            descriptionClass = "noun";
+            check = with (lib.strings);
+              x: path.check x && hasPrefix prefix (toString x);
+            merge = mergeEqualOption;
+          };
+        in
+        submodule {
+          freeformType = attrsOf str;
+          options = {
+            database_path = mkOption {
+              default = stateDirectoryPrefix + "osquery/osquery.db";
+              description = mdDoc "Path used for the database file.";
+              type = pathWithPrefix stateDirectoryPrefix;
+            };
```

Comment:
Do we even need a setting for this? In nixos most if not all data of system services should be under /var/lib

Reply:
I also don't think it useful. However, we should keep this option as they are required in the flag file. I could however set them as `readOnly`.

Reply:
We could also just hardcode the path if we already hide the option.

Reply:
Since this option set is of type `freeform`, having readonly options avoid a user to try to change these option values.